typing/predef.ml

```diff
@@ -163,54 +236,86 @@ let mk_add_type add_type type_ident ?manifest
      type_uid = Uid.of_predef_id type_ident;
     }
   in
-  add_type type_ident decl env
-
-let build_initial_env add_type add_extension empty_env =
-  let add_type = mk_add_type add_type
-  and add_type1 type_ident
-      ~variance ~separability ?(kind=fun _ -> Type_abstract Definition) env =
+  let decl1
+      ~variance
+      ?(separability = Separability.Ind)
+      ?(kind = fun _ -> Type_abstract Definition)
+      type_ident
+    =
     let param = newgenvar () in
-    let decl =
-      {type_params = [param];
-       type_arity = 1;
-       type_kind = kind param;
-       type_loc = Location.none;
-       type_private = Asttypes.Public;
-       type_manifest = None;
-       type_variance = [variance];
-       type_separability = [separability];
-       type_is_newtype = false;
-       type_expansion_scope = lowest_level;
-       type_attributes = [];
-       type_immediate = Unknown;
-       type_unboxed_default = false;
-       type_uid = Uid.of_predef_id type_ident;
-      }
-    in
-    add_type type_ident decl env
-  and add_continuation type_ident env =
-    let tvar1 = newgenvar() in
-    let tvar2 = newgenvar() in
-    let arity = 2 in
-    let decl =
-      {type_params = [tvar1; tvar2];
-       type_arity = arity;
-       type_kind = Type_abstract Definition;
-       type_loc = Location.none;
-       type_private = Asttypes.Public;
-       type_manifest = None;
-       type_variance = [Variance.contravariant; Variance.covariant];
-       type_separability = Types.Separability.default_signature ~arity;
-       type_is_newtype = false;
-       type_expansion_scope = lowest_level;
-       type_attributes = [];
-       type_immediate = Unknown;
-       type_unboxed_default = false;
-       type_uid = Uid.of_predef_id type_ident;
-      }
-    in
-    add_type type_ident decl env
+    { (decl0 ~kind:(kind param) type_ident) with
+      type_params = [param];
+      type_arity = 1;
+      type_variance = [variance];
+      type_separability = [separability];
+    }
   in
+  let decl2
+      ~variance:(var1, var2)
+      ?separability:((sep1, sep2) = (Separability.Ind, Separability.Ind))
+      ?(kind = fun _ _ -> Type_abstract Definition)
+      type_ident
+    =
+    let param1, param2 = newgenvar (), newgenvar () in
+    { (decl0 ~kind:(kind param1 param2) type_ident) with
+      type_params = [param1; param2];
+      type_arity = 2;
+      type_variance = [var1; var2];
+      type_separability = [sep1; sep2];
+    }
+  in
+  let cstr id args =
+    {
+      cd_id = id;
+      cd_args = Cstr_tuple args;
+      cd_res = None;
+      cd_loc = Location.none;
+      cd_attributes = [];
+      cd_uid = Uid.of_predef_id id;
+    }
+  in
+  let variant constrs =
+    Type_variant (constrs, Variant_regular) in
+  function
+  | `Int -> decl0 ~immediate:Always ident_int
+  | `Char -> decl0 ~immediate:Always ident_char
```

Comment:
If the code factored the computation of the associated `ident`, it would reduce the number of location whete this association is defining while allowing to group together predefined types with similar shape: 
```ocaml
fun type_constr ->
  let name = ident_of_type_constr
  | `Int | `Char -> decl0 ~immediate:Always name
  | `String | `Byte | `Float -> decl0 name
 ```

Reply:
I pushed a modified version along those lines. Let me know what you think. I keep it as a separate commit because I think it is easier to review the first version first, which kept the same structure as the previous code, and then this refactoring separately.